src/backend/distributed/utils/enable_ssl.c

```diff
@@ -0,0 +1,429 @@
+		return false;
+	}
+
+	elog(LOG, "no certificate present, generating self signed certificate");
+
+	pkey = generate_key_rsa();
+	if (!pkey)
+	{
+		ereport(ERROR, (errmsg("error while generating private key")));
+
+		return false;
+	}
+
+	x509 = generate_x509(pkey);
+	if (!x509)
+	{
+		ereport(ERROR, (errmsg("error while generating x509 certificate")));
+		return false;
+	}
+
+	/* Write the private key and certificate out to disk. */
+	filesWritten = StoreCertificate(pkey, x509);
+
+	if (!filesWritten)
+	{
+		ereport(ERROR, (errmsg("error while writing key and certificate to file")));
+		return false;
+	}
+
+	return true;
+}
+
+
+/* Generates a 2048-bit RSA key. */
+static EVP_PKEY *
+generate_key_rsa()
+{
+	int ret = 0;
+	EVP_PKEY *pkey = NULL;
+	BIGNUM *bne = NULL;
+	RSA *rsa = NULL;
+
+	/* Allocate memory for the EVP_PKEY structure. */
+	pkey = EVP_PKEY_new();
+	if (!pkey)
+	{
+		ereport(ERROR, (errmsg("unable to create EVP_PLEY")));
+		return NULL;
+	}
+	EnsureFreeOpenSSLResource((MemoryContextCallbackFunction) & EVP_PKEY_free, pkey);
+
+	/* 1. generate rsa key */
+	bne = BN_new();
+	ret = BN_set_word(bne, RSA_F4);
+	if (ret != 1)
+	{
+		ereport(ERROR, (errmsg("unable to prepare bignum")));
+		return NULL;
+	}
+	EnsureFreeOpenSSLResource((MemoryContextCallbackFunction) & BN_free, bne);
+
+	/* Generate the RSA key and assign it to pkey. */
+	rsa = RSA_new();
+	ret = RSA_generate_key_ex(rsa, 2048, bne, NULL);
+	if (ret != 1)
+	{
+		ereport(ERROR, (errmsg("unable to generate RSA key")));
+		return NULL;
+	}
+
+	if (!EVP_PKEY_assign_RSA(pkey, rsa))
+	{
+		ereport(ERROR, (errmsg("unable to assign RSA key")));
+		return NULL;
+	}
+
+	/* The key has been generated, return it. */
+	return pkey;
+}
+
+
+/*
+ * generate_x509 creates a self signed certificate for citus to use. The certificate will
+ * contain the public parts of pkey and will be signed in the end by the private part to
+ * make it self signed.
+ */
+static X509 *
+generate_x509(EVP_PKEY *pkey)
+{
+	X509 *x509 = NULL;
+	X509_NAME *name = NULL;
+
+	x509 = X509_new();
+	if (!x509)
+	{
+		ereport(ERROR, (errmsg("unable to create x509 cert")));
+		return NULL;
+	}
+	EnsureFreeOpenSSLResource((MemoryContextCallbackFunction) & X509_free, x509);
+
+	/* Set the serial number. */
+	ASN1_INTEGER_set(X509_get_serialNumber(x509), 1);
+
+	/*
+	 * Set the expiry of the certificate.
+	 *
+	 * the functions X509_get_notBefore and X509_get_notAfter are deprecated, these are
+	 * replaced with mutable and non-mutable variants in openssl 1.1, however they are
+	 * better supported then the newer versions. In 1.1 they are aliasses to the mutable
+	 * variant (X509_getm_notBefore, ...) that we actually need, so they will actually use
+	 * the correct function in newer versions.
+	 *
+	 * Postgres does not check the validity on the certificates, but we can't omit the
+	 * dates either to create a certificate that can be parsed.
+	 * TODO settle on the actual validity times on the PR
```

Comment:
I'd like some input on the validity of the certificate.

Observation: Postgres seems to not check the validity period of a certificate given that an expiry date set to the creation time does allow the test suite to pass.

Normally a certificate should be rejected after its **not after** time has expired. Currently postgres/libpq does not reject such certificates when the sslmode is set to require. There are roughly 3 options we can choose for

1. Set the expiry date deliberately in the past as currently implemented
2. Set the expiry to a reasonable data (eg. 1 year in the future)
3. Set an extreme expiry date (eg. 100 years in the future)

As long as postgres does not check the validity it doesn't really matter what we put here. So my reasoning to set it to 0 seconds is as follows.

When postgres changes the behaviour of how it checks certificates it would implicitly install a time bomb in the citus installation of a customer as the certificate might still be valid once they upgrade postgres. With a certificate in place that is already expired the user will run into this problem the moment it upgrades postgres to a version that enforces the validity period of the certificate.

If we would install a certificate with a validity of a year it might very well be that postgres gets successfully upgraded to stop working at an unexpected time in the future.

Setting the validity to an extreme expiry effectively negates the change to postgres by allowing the database to accept the certificate for its lifetime.

With the reasoning above I came to the setting of 0 seconds after the issuing time for the expiry. Wonder what other people think about this case.

Reply:
Seems like a good enough reasoning.

Reply:
If we think about how we can secure Citus further, one reason to give it a longer expiration date would be that the user could decide to sign this certificate. In fact, there could be an automated process where the coordinator signs the certificate the first time it connects using a trusted key and sends it back to the worker. In that case we would want a longer expiration date. A year seems somewhat reasonable/standard.

Reply:
When people sign a key they will create a new certificate with their own expiry date. Same goes when the coordinator will become the CA of a citus cluster and signs the workers.

Enough will need to change when we go there that my reasoning above stays for as long as we do self-signed like this.